pkg/flatrpc/conn.go

```diff
@@ -125,14 +109,31 @@ func Send[T sendMsg](c *Conn, msg T) error {
 	return nil
 }
 
-// Recv received an RPC message.
-// The type T is supposed to be a normal flatbuffers type (not ending with T, e.g. ConnectRequest).
+// Recv receives an RPC message.
+// The type T is supposed to be a pointer to a normal flatbuffers type (not ending with T, e.g. *ConnectRequestRaw).
 // Receiving should be done from a single goroutine, the received message is valid
 // only until the next Recv call (messages share the same underlying receive buffer).
-func Recv[T any, PT interface {
-	*T
+func Recv[Raw interface {
+	UnPack() *T
 	flatbuffers.FlatBuffer
-}](c *Conn) (*T, error) {
+}, T any](c *Conn) (res *T, err0 error) {
+	defer func() {
+		if err1 := recover(); err1 != nil {
```

Comment:
mgr.Init and UnPack can panic as follows.
Flatbuffers assumes correct well-formed data. C++ has separate verification function, but Go does. And it's still safer and faster to just catch all panics. We can get bad messages if VM memory is corrupted, or due to bugs in executor.


```
panic: runtime error: slice bounds out of range [:3402220082] with capacity 221180

goroutine 15154 [running]:
[github.com/google/flatbuffers/go.(*Table).ByteVector(0x7fa62a66b5b8](http://github.com/google/flatbuffers/go.(*Table).ByteVector(0x7fa62a66b5b8)?, 0x40?)
/syzkaller/gopath/src/[github.com/google/syzkaller/vendor/github.com/google/flatbuffers/go/table.go:38](http://github.com/google/syzkaller/vendor/github.com/google/flatbuffers/go/table.go:38) +0xaa
[github.com/google/syzkaller/pkg/flatrpc.(*ExecResultRaw).Error(0xc0037a3c40)](http://github.com/google/syzkaller/pkg/flatrpc.(*ExecResultRaw).Error(0xc0037a3c40))
/syzkaller/gopath/src/[github.com/google/syzkaller/pkg/flatrpc/flatrpc.go:2945](http://github.com/google/syzkaller/pkg/flatrpc/flatrpc.go:2945) +0x36
[github.com/google/syzkaller/pkg/flatrpc.(*ExecResultRaw).UnPackTo(0x12388](http://github.com/google/syzkaller/pkg/flatrpc.(*ExecResultRaw).UnPackTo(0x12388)?, 0xc008014680)
/syzkaller/gopath/src/[github.com/google/syzkaller/pkg/flatrpc/flatrpc.go:2856](http://github.com/google/syzkaller/pkg/flatrpc/flatrpc.go:2856) +0x6b
[github.com/google/syzkaller/pkg/flatrpc.(*ExecResultRaw).UnPack(..](http://github.com/google/syzkaller/pkg/flatrpc.(*ExecResultRaw).UnPack(..).)
/syzkaller/gopath/src/[github.com/google/syzkaller/pkg/flatrpc/flatrpc.go:2865](http://github.com/google/syzkaller/pkg/flatrpc/flatrpc.go:2865)
[github.com/google/syzkaller/pkg/flatrpc.ExecutorMessagesRaw.UnPack(0xc0](http://github.com/google/syzkaller/pkg/flatrpc.ExecutorMessagesRaw.UnPack(0xc0)?, {{0xc008f7c004?, 0x0?, 0x0?}, 0x0?})
/syzkaller/gopath/src/[github.com/google/syzkaller/pkg/flatrpc/flatrpc.go:203](http://github.com/google/syzkaller/pkg/flatrpc/flatrpc.go:203) +0xd7
[github.com/google/syzkaller/pkg/flatrpc.(*ExecutorMessageRaw).UnPackTo(0x1923f80](http://github.com/google/syzkaller/pkg/flatrpc.(*ExecutorMessageRaw).UnPackTo(0x1923f80)?, 0xc0037a3d68)
/syzkaller/gopath/src/[github.com/google/syzkaller/pkg/flatrpc/flatrpc.go:1602](http://github.com/google/syzkaller/pkg/flatrpc/flatrpc.go:1602) +0x66
[github.com/google/syzkaller/pkg/flatrpc.(*ExecutorMessageRaw).UnPack(..](http://github.com/google/syzkaller/pkg/flatrpc.(*ExecutorMessageRaw).UnPack(..).)
/syzkaller/gopath/src/[github.com/google/syzkaller/pkg/flatrpc/flatrpc.go:1611](http://github.com/google/syzkaller/pkg/flatrpc/flatrpc.go:1611)
main.(*RPCServer).connectionLoop(0xc0017306c0, 0xc00157c000)
/syzkaller/gopath/src/[github.com/google/syzkaller/syz-manager/rpc.go:301](http://github.com/google/syzkaller/syz-manager/rpc.go:301) +0x305
main.(*RPCServer).handleConn(0xc0017306c0, 0xc003d66000)
/syzkaller/gopath/src/[github.com/google/syzkaller/syz-manager/rpc.go:170](http://github.com/google/syzkaller/syz-manager/rpc.go:170) +0x43d
[github.com/google/syzkaller/pkg/flatrpc.ListenAndServe.func1.1()](http://github.com/google/syzkaller/pkg/flatrpc.ListenAndServe.func1.1())
/syzkaller/gopath/src/[github.com/google/syzkaller/pkg/flatrpc/conn.go:55](http://github.com/google/syzkaller/pkg/flatrpc/conn.go:55) +0x122
created by [github.com/google/syzkaller/pkg/flatrpc.ListenAndServe.func1](http://github.com/google/syzkaller/pkg/flatrpc.ListenAndServe.func1) in goroutine 15
/syzkaller/gopath/src/[github.com/google/syzkaller/pkg/flatrpc/conn.go:52](http://github.com/google/syzkaller/pkg/flatrpc/conn.go:52) +0xb1
```

Reply:
Is there a chance that it will not (just) panic, but also corrupt other data/memory?

Reply:
No, it shouldn't. There are no major uses of unsafe in flatbuffers, and otherwise Go is memory safe (data races aside):

```
syzkaller💻 git grep unsafe vendor/github.com/google/flatbuffers/
vendor/github.com/google/flatbuffers/go/sizes.go:       "unsafe"
vendor/github.com/google/flatbuffers/go/sizes.go:       return *(*string)(unsafe.Pointer(&b))
```